ssl/ssl_transfer_asn1.cc

```diff
@@ -717,20 +811,28 @@ static int SSL3_STATE_from_bytes(SSL *ssl, CBS *cbs, const SSL_CTX *ctx) {
       return 0;
     }
     CBS app_seq;
-    uint64_t pending_app_data_offset, pending_app_data_size;
+    int64_t pending_app_data_offset;
+    uint64_t pending_app_data_size;
     if (!CBS_get_asn1(&pending_app_data, &app_seq, CBS_ASN1_SEQUENCE) ||
-        !CBS_get_asn1_uint64(&app_seq, &pending_app_data_offset) ||
+        !CBS_get_asn1_int64(&app_seq, &pending_app_data_offset) ||
         !CBS_get_asn1_uint64(&app_seq, &pending_app_data_size)) {
       OPENSSL_PUT_ERROR(SSL, SSL_R_SERIALIZATION_INVALID_SSL3_STATE);
       return 0;
     }
-    if (pending_app_data_size > out->read_buffer.buf_size()) {
+    if (!fits_in_ptrdiff_int64(pending_app_data_offset)) {
+      OPENSSL_PUT_ERROR(SSL, SSL_R_SERIALIZATION_INVALID_SSL3_STATE);
+      return 0;
+    }
+    uint8_t *pending_app_data_ptr =
+        out->read_buffer.data() + (ptrdiff_t)pending_app_data_offset;
+    if (out->read_buffer.buf_ptr() > pending_app_data_ptr ||
+        pending_app_data_ptr + pending_app_data_size >
+            out->read_buffer.buf_ptr() + out->read_buffer.buf_size()) {
       OPENSSL_PUT_ERROR(SSL, SSL_R_SERIALIZATION_INVALID_SSL3_STATE);
       return 0;
     }
     out->pending_app_data =
-        MakeSpan(out->read_buffer.buf_ptr() + pending_app_data_offset,
-                 pending_app_data_size);
+        MakeSpan(pending_app_data_ptr, pending_app_data_size);
   }
```

Comment:
On a fresh re-read, I realize this weeds to special case the previous format version behavior from the new behavior.